text/trace/0157-messaging-semantic-conventions.md

```diff
@@ -0,0 +1,210 @@
+   processed, or it is settled automatically when it is received.
+
+The semantic conventions need to define how to handle failures and retries in
+all stages that interface with the intermediary: publish, receive and settle.
+
+Based on this model, the following scenarios capture major requirements and
+can be used for prototyping, as examples, and as test cases.
+
+### Individual settlement
+
+Individual settlement systems imply independent logical message flows. A single
+message is created and published in the same context, and it's delivered,
+consumed, and settled as a single entity. Each message needs to be settled
+individually.
+
+Transport batching can be treated as a special case: messages can be
+transported together as an optimization, but are produced and consumed
+individually.
+
+```
++---------+ +---------+ +---------+ +---------+ +---------+ +---------+
+|Message A| |Message B| |Message C| |Message D| |Message E| |Message F|
++---------+ +---------+ +---------+ +---------+ +---------+ +---------+
+ Settled                 Settled                             Settled 
+```
+
+#### Examples
+
+1. RabbitMQ: TODO
+
+### Checkpoint-based settlement
+
+Messages are processed as a stream and settled to specific checkpoints. A
+checkpoint points to a position of the stream up to which messages were
+processed and settled. Messages cannot be settled individually, instead, the
+checkpoint needs to be forwarded.
+
+Checkpoint-based settlement systems are designed to efficiently receive and
+settle batches of messages. However, it is not possible to settle messages
+independent of their position in the stream (e. g., if message B is located at
+a later position in the stream than message A, then message B cannot be settled
+without also settling message A).
+
+```
+                               Checkpoint
+                                   | 
+                                   v                                
++---------+ +---------+ +---------+ +---------+ +---------+ +---------+
+|Message A| |Message B| |Message C| |Message D| |Message E| |Message F|
++---------+ +---------+ +---------+ +---------+ +---------+ +---------+
+                     <---  Settled
+```
+
+#### Examples
+
+1. The following configurations should be instrumented and tested for Kafka or
+   a similar messaging system:
+
+   * 1 producer, 2 consumers in the same consumer group
+   * 1 producer, 2 consumers in different consumer groups
+   * 2 producers, 2 consumers in the same consumer group
+
+   Each of the producers produces a continuous stream of messages.
+
+## Open questions
+
+The following areas are considered out-of-scope of a first stable release of
+semantic conventions for messaging. While not being explicitly considered for
+a first stable release, it is important to ensure that this first stable
+release can serve as a solid foundation for further improvements in these areas.
+
```

Comment:
Can we briefly talk about spec boundaries? E.g. if the application produces/consumes/routes batches of 100x messages, tracing them might not be a great option because of the performance, storage needs of expressing 100x of links, and gantt-charts not being really useful.

Reply:
We definitely need to talk about that, but I'd prefer to do so once we agreed on a set of scenarios. Reasoning about boundaries already requires us to consider and evaluate possible solutions.

Reply:
I added this to the open question about sampling, making clear that link-based sampling is out of scope, but that questions about the number of links in a single trace need to be discussed and considered by the conventions:
> However, having a too high number of span links in a single trace or having too many traces linked together can make the visualization and analysis of traces inefficient. This problem is not related to sampling and needs to be addressed by the semantic conventions.